### PR TITLE
[WIP] Add Phoenix coder

### PR DESCRIPTION
**Pull request by weiqingy**

### CODE REVIEW

**Comment by weiqingy** [comment]
Code Change:

- Refactor the code of date type conversions in `Utils.scala` into corresponding data type classes (Avro, PrimitiveType, Phoenix). For composite key, only PrimitiveType support it in this PR.
- Except composite key, add Phoenix data type to do encoder/decoder.

How to test:
Pass current unit tests and add new unit tests.

Hi, @dongjoon-hyun @merlintang @bikassaha Would you please help to review this PR? Thanks.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseFilter.scala [hunk 1]
```diff
@@ -183,14 +179,14 @@ object HBaseFilter extends Logging{
           val f = relation.getField(attribute)
           val filter = bound.greater.map { x =>
             val lower = new SingleColumnValueFilter(
-              Bytes.toBytes(f.cf),
-              Bytes.toBytes(f.col),
+              SHCDataTypeFactory.create(tCoder).toBytes(f.cf),
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseFilter.scala]: Can we move this `SHCDataTypeFactory.create(tCoder)` into line 163 like `tCoder`?

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Phoenix.scala [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      case data: Boolean => PBoolean.INSTANCE.toBytes(data)
+      case data: Byte => PTinyint.INSTANCE.toBytes(data)
+      case data: Array[Byte] => data
+      case data: Double => PDouble.INSTANCE.toBytes(data)
+      case data: Float => PFloat.INSTANCE.toBytes(data)
+      case data: Int => PInteger.INSTANCE.toBytes(data)
+      case data: Long => PLong.INSTANCE.toBytes(data)
+      case data: Short => PSmallint.INSTANCE.toBytes(data)
+      case data: String => PVarchar.INSTANCE.toBytes(data)
+      case _ => throw new Exception(s"unsupported data type $input")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Phoenix.scala]: `UnsupportedOperationException` instead of `Exception`?

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Phoenix.scala [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      case BooleanType => PBoolean.INSTANCE
+      case ByteType => PTinyint.INSTANCE
+      case DoubleType => PDouble.INSTANCE
+      case IntegerType => PInteger.INSTANCE
+      case FloatType => PFloat.INSTANCE
+      case LongType => PLong.INSTANCE
+      case ShortType => PSmallint.INSTANCE
+      case StringType => PVarchar.INSTANCE
+      case BinaryType => PBinary.INSTANCE
+      case _ => throw new Exception(s"unsupported data type $input")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Phoenix.scala]: `UnsupportedOperationException` instead of `Exception`?

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/PrimitiveType.scala [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      case data: Byte => Array(data)
+      case data: Array[Byte] => data
+      case data: Double => Bytes.toBytes(data)
+      case data: Float => Bytes.toBytes(data)
+      case data: Int => Bytes.toBytes(data)
+      case data: Long => Bytes.toBytes(data)
+      case data: Short => Bytes.toBytes(data)
+      case data: UTF8String => data.getBytes
+      case data: String => Bytes.toBytes(data)
+      case _ => throw new Exception(s"unsupported data type $input")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/PrimitiveType.scala]: `UnsupportedOperationException` instead of `Exception`?

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Avro.scala [hunk 5]
```diff
@@ -377,22 +318,108 @@ object SchemaConverters {
...
+            s"This mix of union types is not supported (see README): $other")
+        }
+      case other => throw new SchemaConversionException(s"invalid avro type: $other")
+    }
+  }
 }
 
 
-object AvroSedes {
+object AvroSerde {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Avro.scala]: Is this renaming necessary? This seems to make many meaningless changes.
If possible, let's not change this.

File: core/src/test/scala/org/apache/spark/sql/HBaseTestSuite.scala [hunk 6]
```diff
@@ -19,13 +19,14 @@ package org.apache.spark.sql
 
 import java.io.File
 
+import org.scalatest.{BeforeAndAfterAll, BeforeAndAfterEach, FunSuite}
+import scala.collection.JavaConverters._
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/sql/HBaseTestSuite.scala]: Is this the result of `ScalaStyle` check?

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseFilter.scala (hunk 1 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseFilter.scala]: Yes. I have updated the PR.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Phoenix.scala (hunk 2 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Phoenix.scala]: yes.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Phoenix.scala (hunk 3 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Phoenix.scala]: Yes.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/PrimitiveType.scala (hunk 4 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/PrimitiveType.scala]: Yes.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Avro.scala (hunk 5 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Avro.scala]: Actually, the short name of `serialize/deserialize` used widely is `serde` instead of  `sedes`, so I renamed it. (it might be a spelling error.)

File: core/src/test/scala/org/apache/spark/sql/HBaseTestSuite.scala (hunk 6 above)
Comment (weiqingy) [inline core/src/test/scala/org/apache/spark/sql/HBaseTestSuite.scala]: I make the order of `import` in SHC like this: java -> scala -> apache -> internal class. What do you think about the order?

**Comment by weiqingy** [comment]
Hi, @dongjoon-hyun  Yes. I will pay attention to make the PR easier to review next time. Sorry about this. I have updated the PR. Could you please review it again? Thanks.

File: core/src/test/scala/org/apache/spark/sql/HBaseTestSuite.scala (hunk 6 above)
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/sql/HBaseTestSuite.scala]: +1 for the order. What I wanted is we need here some automation to be consistent.

**Comment by dongjoon-hyun** [comment]
Yes, @weiqingy !

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Avro.scala (hunk 5 above)
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Avro.scala]: Yep. I think so. But, this kind of typos having many usages should be handled separate PR.
Could you make a separate PR for that only? I'll merge that first.

Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Avro.scala]: If that code is identical, `rebasing` will be enough for this PR after merging that renaming PR.

File: core/src/test/scala/org/apache/spark/sql/HBaseTestSuite.scala (hunk 6 above)
Comment (weiqingy) [inline core/src/test/scala/org/apache/spark/sql/HBaseTestSuite.scala]: Yes, I have add an issue to record this: https://github.com/hortonworks-spark/shc/issues/80.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Avro.scala (hunk 5 above)
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Avro.scala]: If that is difficult, please ignore this~

Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/Avro.scala]: Sure. I am going to submit a separate PR for the renaming only.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  def encodeCompositeRowKey(rkIdxedFields:Seq[(Int, Field)], row: Row): Seq[Array[Byte]]
+}
+
+/**
+ * Currently, SHC supports three data types which can be used as serdes: Avro, Phoenix, PrimitiveType.
+ * Adding New SHC data type should needs to implement the trait 'SHCDataType'.
+ */
+object SHCDataTypeFactory {
+
+  def create(f: Field): SHCDataType = {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala]: Can we implement this by using `def create(coder: String)`?

Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala]: Could you add a guard code to check `f is null`?

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+      Class.forName(f.fCoder)
+        .getConstructor(classOf[Option[Field]])
+        .newInstance(f.fCoder)
+        .asInstanceOf[SHCDataType]
+    }
+  }
+
+  def create(coder: String): SHCDataType = {
+    if (coder == classOf[Avro].getSimpleName) {
+      new Avro()
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala]: Can we reuse a single instance and return that here?

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala (hunk 7 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala]: I think we have to use `Field` instead of `string` here since some places like [here](https://github.com/hortonworks-spark/shc/pull/79/files#diff-29979b28b7592cf3830571505ba73d9dR44) and [here](https://github.com/hortonworks-spark/shc/pull/79/files#diff-2ddaf40e93f41d594241a44e468d686bR31) needs the information insides `Field` to do encode/decode. For the cases which do not need the information inside `Field` like [here](https://github.com/hortonworks-spark/shc/pull/79/files#diff-2ddaf40e93f41d594241a44e468d686bR50), shc will use another `def create(coder: String): SHCDataType = {` funtion to new the `SHCDataType`.

Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala]: The guard code is inside each data type. For example, in `class PrimitiveType`, the guard code is [here](https://github.com/hortonworks-spark/shc/pull/79/files#diff-2ddaf40e93f41d594241a44e468d686bR44).

Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala]: It's not the same. The guard you mentioned is used after `SHCDataTypeFactory.create(..)` succeed. The guard here will prevent NPE in `SHCDataTypeFactory.create()`.

**Comment by dongjoon-hyun** [comment]
Sorry for the conflicts. I merged that. Could you rebase this PR?

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala (hunk 7 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala]: I see. Yep, I will update the PR.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala (hunk 8 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala]: Yes. I will update the PR. We can reuse a single instance here.

**Comment by weiqingy** [comment]
Thanks, @dongjoon-hyun . I am going to update the PR.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala [hunk 9]
```diff
@@ -292,98 +294,151 @@ case class BoundRange(
...
+            Array(BoundRange(Bytes.toBytes(0: Int), Bytes.toBytes(Integer.MAX_VALUE)),
+              BoundRange(b, Bytes.toBytes(-1: Integer))),
+            b))
+        }
       } else {
-        Some(BoundRanges(
-          Array(BoundRange(Bytes.toBytes(Integer.MIN_VALUE), b)),
-          Array(BoundRange(Bytes.toBytes(0: Int), Bytes.toBytes(Integer.MAX_VALUE)),
-            BoundRange(b, Bytes.toBytes(-1: Integer))), b))
+        // Non-PrimitiveType
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: What about the following?
```scala
val f = SHCDataTypeFactory.create(f)
val b = f.toBytes(in)
...
```

Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: In this file, there is several occurrence to call redundant call.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala [hunk 10]
```diff
@@ -292,98 +294,151 @@ case class BoundRange(
...
+            Array(BoundRange(Bytes.toBytes(Long.MinValue), b)),
+            Array(BoundRange(Bytes.toBytes(0: Long), Bytes.toBytes(Long.MaxValue)),
+              BoundRange(b, Bytes.toBytes(-1: Long))), b))
+        }
       } else {
-        Some(BoundRanges(
-          Array(BoundRange(Bytes.toBytes(Long.MinValue), b)),
-          Array(BoundRange(Bytes.toBytes(0: Long), Bytes.toBytes(Long.MaxValue)),
-            BoundRange(b, Bytes.toBytes(-1: Long))), b))
+        // Non-PrimitiveType
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: Here.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala [hunk 11]
```diff
@@ -292,98 +294,151 @@ case class BoundRange(
...
+            Array(BoundRange(Bytes.toBytes(Short.MinValue), b)),
+            Array(BoundRange(Bytes.toBytes(0: Short), Bytes.toBytes(Short.MaxValue)),
+              BoundRange(b, Bytes.toBytes(-1: Short))), b))
+        }
       } else {
-        Some(BoundRanges(
-          Array(BoundRange(Bytes.toBytes(Short.MinValue), b)),
-          Array(BoundRange(Bytes.toBytes(0: Short), Bytes.toBytes(Short.MaxValue)),
-            BoundRange(b, Bytes.toBytes(-1: Short))), b))
+        // Non-PrimitiveType
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: Here.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala [hunk 12]
```diff
@@ -292,98 +294,151 @@ case class BoundRange(
...
+            Array(BoundRange(b, Bytes.toBytes(Double.MinValue))),
+            Array(BoundRange(Bytes.toBytes(0.0d), Bytes.toBytes(Double.MaxValue)),
+              BoundRange(Bytes.toBytes(-0.0d), b)), b))
+        }
       } else {
-        Some(BoundRanges(
-          Array(BoundRange(b, Bytes.toBytes(Double.MinValue))),
-          Array(BoundRange(Bytes.toBytes(0.0d), Bytes.toBytes(Double.MaxValue)),
-            BoundRange(Bytes.toBytes(-0.0d), b)), b))
+        // Non-PrimitiveType
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: Here.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala [hunk 13]
```diff
@@ -292,98 +294,151 @@ case class BoundRange(
...
+            Array(BoundRange(b, Bytes.toBytes(Float.MinValue))),
+            Array( BoundRange(Bytes.toBytes(0.0f), Bytes.toBytes(Float.MaxValue)),
+              BoundRange(Bytes.toBytes(-0.0f), b)), b))
+        }
       } else {
-        Some(BoundRanges(
-          Array(BoundRange(b, Bytes.toBytes(Float.MinValue))),
-          Array( BoundRange(Bytes.toBytes(0.0f), Bytes.toBytes(Float.MaxValue)),
-            BoundRange(Bytes.toBytes(-0.0f), b)), b))
+        // Non-PrimitiveType
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: Here.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala [hunk 14]
```diff
@@ -292,98 +294,151 @@ case class BoundRange(
 case class BoundRanges(less: Array[BoundRange], greater: Array[BoundRange], value: Array[Byte])
 
 object BoundRange extends Logging{
-  def apply(in: Any): Option[BoundRanges] = in match {
+
+  val pt = classOf[PrimitiveType].getSimpleName
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: Could we make a constant for these?
```
classOf[PrimitiveType].getSimpleName
```

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala (hunk 8 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala]: I want to double confirm with you, @dongjoon-hyun . I just remember why I did not reuse a single instance here is:
If we reuse a single instance, the object `SHCDataTypeFactory` will alway hold 4 member variables ( three `SHCDataType`(Avro, Phoenix, PrimitiveType) and a list for the data types plugged into SHC by users), and need to search every time (but we can just ignore this cost). But, if we always create a new SHCDataType which does not have any member (as there is no parameter), so I think the creating cost is not that much. Also, SHCDataTypeFactory does not need to hold anything. That is why I chose to create the instance every time in this PR. What do you think about this?

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+trait SHCDataType {
+  // Parse the hbase Field to it's corresponding Scala type which can then be put into
+  // a Spark GenericRow which is then automatically converted by Spark.
+  def fromBytes(src: HBaseType): Any
+
+  // Convert input to Byte Array (HBaseType)
+  def toBytes(input: Any): Array[Byte]
+
+  // If your data type do not need to support composite keys, you can just leave it empty or
+  // threw an exception to remind users composite key is not supported.
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala]: Let's have default implementation here

```
def isCompositeKeySupported(): Boolean = false

def decodeCompositeRowKey(src: HBaseType, offset: Int, length: Int): Any = {
   throw new UnsupportedOperationException("Composite key is not supported")
}

def encodeCompositeRowKey(rkIdxedFields: Seq[(Int, Field)], row: Row): Seq[Array[Byte]] = {
   throw new UnsupportedOperationException("Composite key is not supported")
}
```

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseRelation.scala [hunk 16]
```diff
@@ -121,8 +122,8 @@ case class HBaseRelation(
           logDebug(s"add family $x to ${catalog.name}")
           tableDesc.addFamily(cf)
         }
-        val startKey = Bytes.toBytes("aaaaaaa")
-        val endKey = Bytes.toBytes("zzzzzzz")
+        val startKey = SHCDataTypeFactory.create(catalog.tCoder).toBytes("aaaaaaa")
+        val endKey = SHCDataTypeFactory.create(catalog.tCoder).toBytes("zzzzzzz")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseRelation.scala]:
```scala
val tCoder = SHCDataTypeFactory.create(catalog.tCoder)
val startKey = tCoder.toBytes("aaaaaaa")
val endKey = tCoder.toBytes("zzzzzzz")
```

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala [hunk 17]
```diff
@@ -136,22 +137,46 @@ case class HBaseTableCatalog(
...
-  initRowKey
+  initRowKey()
+
+  def validateCatalogDef() = {
+    if (tCoder == classOf[Avro].getSimpleName) {
+      throw new UnsupportedOperationException("Avro can only be column's coder")
+    }
+
+    if (coderSet.size > 1){
+      // Only Avro can be used with anther one coder
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: `anther one coder` -> `another coder`

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableScan.scala [hunk 18]
```diff
@@ -88,20 +88,20 @@ private[hbase] class HBaseTableScanRDD(
...
    * @param keyFields all of the fields in the row key, ORDERED by their order in the row key.
    */
-  def parseRowKey(row: Array[Byte], keyFields: Seq[Field]): Map[Field, Any] = {
+  def parseCompositeRowKey(row: Array[Byte], keyFields: Seq[Field]): Map[Field, Any] = {
     keyFields.foldLeft((0, Seq[(Field, Any)]()))((state, field) => {
       val idx = state._1
       val parsed = state._2
       if (field.length != -1) {
-        val value = Utils.hbaseFieldToScalaType(field, row, idx, field.length)
+        val value = SHCDataTypeFactory.create(field).decodeCompositeRowKey(row, idx, field.length)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableScan.scala]: Let's move `SHCDataTypeFactory.create(field).` into outside of `if`.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableScan.scala [hunk 19]
```diff
@@ -88,20 +88,20 @@ private[hbase] class HBaseTableScanRDD(
...
         // Return the new index and appended value
         (idx + field.length, parsed ++ Seq((field, value)))
       } else {
         // This is the last dimension.
-        val value = Utils.hbaseFieldToScalaType(field, row, idx, row.length - idx)
+        val value = SHCDataTypeFactory.create(field).decodeCompositeRowKey(row, idx, row.length - idx)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableScan.scala]: This one together.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala (hunk 9 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: Yes.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala (hunk 10 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: Yes.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala (hunk 11 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: Yes.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala (hunk 12 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: Yes.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala (hunk 13 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: Yes.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala (hunk 8 above)
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala]: Creating object will make GC. How many time do you think `create` will be called?

File: core/src/test/scala/org/apache/spark/sql/Test.scala [hunk 20]
```diff
@@ -23,7 +23,12 @@ import org.apache.spark.sql.types.BinaryType
 
 object Test {
   def main(args: Array[String]) {
-   val a: Array[Byte] = Array.fill(10)(Byte.MinValue)
+    val ret1 = Bytes.toBytes(100000L)
+    val len0 = ret1.length.toShort
+    val ret2 = Bytes.toBytes(len0)
+    val ret3 = ret2 ++ ret1
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/sql/Test.scala]: What is the test purpose of the above 4 lines?

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala (hunk 14 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: I am natural for this. Actually, I did this in the previous PR which make constants for these in [object SparkHBaseConf](https://github.com/hortonworks-spark/shc/blob/master/core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/SparkHBaseConf.scala#L23) . But comparing with using a static string to recode it, I think classOf[Avro].getSimpleName is more readable, although there is some minor cost.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala (hunk 15 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/types/SHCDataType.scala]: Yes.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseRelation.scala (hunk 16 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseRelation.scala]: Yes.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala (hunk 17 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: Yes. :)

File: core/src/test/scala/org/apache/spark/sql/Test.scala (hunk 20 above)
Comment (weiqingy) [inline core/src/test/scala/org/apache/spark/sql/Test.scala]: No purpose. All the tests inside Test.scala, I think all of them are meaningless, just some tests  created during the developing phase. I can remove my changes in it.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala (hunk 14 above)
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/ScanRange.scala]: This consists of two function calls: Scala `classOf` function and `getSimpleName` function.
Function call cannot be minor if you use like the following.
```scala
  def create(f: Field): SHCDataType = {
    if (f.fCoder == classOf[Avro].getSimpleName) {
      new Avro(Some(f))
    } else if (f.fCoder == classOf[Phoenix].getSimpleName) {
      new Phoenix(Some(f))
    } else if (f.fCoder == classOf[PrimitiveType].getSimpleName) {
      new PrimitiveType(Some(f))
    } else {
      // Data type implemented by user
      Class.forName(f.fCoder)
        .getConstructor(classOf[Option[Field]])
        .newInstance(f.fCoder)
        .asInstanceOf[SHCDataType]
    }
  }
```

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableScan.scala (hunk 18 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableScan.scala]: Yes.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableScan.scala (hunk 19 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableScan.scala]: Yes.

File: core/src/test/scala/org/apache/spark/sql/Test.scala (hunk 20 above)
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/sql/Test.scala]: Then, let's revert all changes on this file.

**Comment by dongjoon-hyun** [comment]
@weiqingy . I finished the second round.

**Comment by weiqingy** [comment]
Hi, @dongjoon-hyun I have updated the PR based on your comments. Could you please review it again? Thanks.

**Comment by dongjoon-hyun** [comment]
LGTM.
cc @bikassaha .

**Comment by weiqingy** [comment]
Hi, @dongjoon-hyun Based on our discussion, I have updated the PR to make the table coder creation only happen once. Could you please check it (the commit [75a3620](https://github.com/hortonworks-spark/shc/pull/79/commits/75a36207d52a13930bb7d10f7ca83a76b1c1b23a)) again? Thanks.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala [hunk 21]
```diff
@@ -137,6 +137,8 @@ case class HBaseTableCatalog(
     sMap.fields.map(_.cf).filter(_ != HBaseTableCatalog.rowKey).toSeq.distinct
   }
 
+  val shcTableCoder = SHCDataTypeFactory.create(tCoder)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseTableCatalog.scala]: Thanks. Much better!

**Comment by dongjoon-hyun** [comment]
@bikassaha and @weiqingy .
Let's merge this and move forward! :)

**Comment by weiqingy** [comment]
Thanks, @dongjoon-hyun . Merged this to Master branch.
